File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    private static ArrayDeque<Node> queue;
    private static boolean[][] visited;
    
    public static class Node {
        int x;
        int y;
        int count;
        
        public Node(int x, int y, int count) {
            this.x = x;
            this.y = y;
            this.count = count;
        }
    }
    
    private static int bfs(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        queue = new ArrayDeque<>();
        visited = new boolean[n][m];
        
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        
        Node node = new Node(0, 0, 1);
        
        queue.addLast(node);
        
        visited[0][0] = true;
        
        while (!queue.isEmpty()) {
            Node cur = queue.pollFirst();
            int curX = cur.x;
            int curY = cur.y;
            int count = cur.count;
            
            if (curX == n - 1 && curY == m - 1) return count;
            
            for (int i = 0; i < 4; i++) {
                int nextX = curX + dx[i];
                int nextY = curY + dy[i];
                
                if (nextX >= 0 && nextY >= 0 && nextX < n && nextY < m) {
                    if (!visited[nextX][nextY] && maps[nextX][nextY] == 1) {
                        visited[nextX][nextY] = true;
                        queue.add(new Node(nextX, nextY, count + 1));
                    }
                }
            }
        }
        return -1;
    }
        
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        answer = bfs(maps);
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = -1

    n = len(maps)
    m = len(maps[0])

    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    visited = set()
    queue = deque()
    
    visited.add((0,0))
    queue.append((0, 0, 1))
    
    while queue:
        x, y, cnt = queue.popleft()
        
        if (x, y) == (n - 1, m - 1):
            return cnt
        
        for i in range(4):
            next_x = x + dx[i]
            next_y = y + dy[i]
            if 0 <= next_x < n and 0 <= next_y < m and maps[next_x][next_y] == 1:
                
                if (next_x, next_y) not in visited:
                    visited.add((next_x, next_y))
                    queue.append([next_x, next_y, cnt + 1])
        
    return answer